PLEP-0007.rst

```diff
@@ -0,0 +1,847 @@
+inference and multiple dispatch to achieve performance comparable to C
+and Fortran. Unlike C and Fortran, Julia can be run interactively and
+does not require type declarations. These features greatly speed up
+code development by allowing prototyping in the same language to be
+used for performance runs. Julia natively supports parallelization,
+and has been used to achieve petascale computing. **Julia proves that
+high performance can be achieved with a dynamically typed interactive
+language without sacrificing usability.**
+
+Julia can call code from Fortran and C, and can act as a wrapper for
+codes written in these compiled languages. The main drawback of this
+approach is that the resulting code would be harder to maintain
+because developers would need to know two or three languages. A
+potential drawback is that problems can arise in practice when code in
+one language is called from a different language. Global optimizations
+might also not be possible when mixing more than one language (though
+Julia may become capable of optimizations across language boundaries
+in the future. If possible, the package itself should be written
+entirely in Julia and depend only on packages that can be installed
+using Julia's built-in package manager. Users shall *not* be required
+to compile or install any external libraries or use any shell scripts.
+
+The implementation shall be written entirely in Julia, and shall have
+a Python interface in addition to a Julia interface. The interface may
+either be included in the PlasmaPy core package or in an affiliated
+package. The ``Plasma`` class should be able to handle the output of
+simulations performed using this plasma simulation framework.
+
+Implementation Notes
+====================
+
+Boundary Conditions
+-------------------
+
+A goal of this effort is to make the setup of fluid, particle, and
+hybrid simulations as similar as possible. Ideally, the same problem
+setup object should be able to be used to initialize all of these
+different types of simulations as similarly as possible. However, the
+formulation of boundary conditions between fluid and PIC simulations
+can be substantially different and potentially incompatible.
+
+.. I'm not sure how to handle this yet, particularly because I do not
+   know enough about boundary conditions for PIC simulations. -Nick
+
+Abstract Interfaces
+-------------------
+
+`Abstract base classes
+<https://docs.python.org/3.7/library/abc.html>`_ (ABCs) in Python
+allow users to define what methods and attributes must be defined in a
+subclass of that ABC. This functionality is used in PlasmaPy's
+``Plasma`` class. An equivalent to ABCs has not yet been implemented
+in Julia (see `Julia issue #6875 on GitHub
+<https://github.com/JuliaLang/julia/issues/6975>`_). An alternative to
+ABCs would be to create a macro that checks that a particular class or
+class instance has all of the required methods.
```

Comment:
I'm not even 100% sure we need ABCs. Simple classes that don't check whether the subclass implements every method but implements what it does with a common API, with a well-designed inheritance pattern for the base classes, could also do the trick.